File: backend/src/domain/users/account-rules.ts
import { ADMIN_ROLES, VALID_ROLES } from './roles';

/**
 * The rules that stop an administrator from locking themselves out.
 *
 * They live here, pure, because they are rules and not queries: each describes a situation
 * where the UI would happily accept an action nobody could come back from. There is no
 * rescue screen in Zanshin — no remaining
 * administrateur actif, et il faut une session psql.
 */

export const MINIMUM_PASSWORD_LENGTH = 12;

const USERNAME = /^[A-Za-z0-9._-]{2,64}$/;

export function isAdminRole(role: string): boolean {
    return (ADMIN_ROLES as readonly string[]).includes(role);
}

/** `null` if the name is acceptable, otherwise the message to show. */
export function validateUsername(username: string): string | null {
    if (!username) return "L'identifiant est requis.";
    if (!USERNAME.test(username)) {
        return 'Invalid username: 2 to 64 characters, letters, digits, ". _ -".';
    }
    return null;
}

/**
 * `null` si le mot de passe est acceptable.
 *
 * A minimum length and nothing else — no composition rule. Character-class requirements
 * produce `Password1!` and encourage reuse; length is the only constraint whose effect on
 * entropy is real.
 */
export function validatePassword(password: string): string | null {
    if (!password) return 'Le mot de passe est requis.';
    if (password.length < MINIMUM_PASSWORD_LENGTH) {
        return `The password must be at least ${MINIMUM_PASSWORD_LENGTH} characters.`;
    }
    // bcrypt truncates at 72 bytes: past that, the extra characters protect nothing, and
    // letting someone believe otherwise would be worse than refusing.
    if (Buffer.byteLength(password, 'utf8') > 72) {
        return 'The password exceeds 72 bytes, past which bcrypt ignores the rest.';
    }
    return null;
}

export function validateRole(role: string): string | null {
    return (VALID_ROLES as readonly string[]).includes(role)
        ? null
        : `Unknown role "${role}". Expected ${VALID_ROLES.join(', ')}.`;
}

/**
 * `null` if the change is allowed, otherwise why it is refused.
 *
 * `remainingActiveAdmins` compte les administrateurs actifs **autres que celui-ci**.
 * Refused at the account level rather than the screen's: three tabs open on two accounts
 * would otherwise be enough to empty the administrator list.
 */
export function refuseSelfLockout(options: {
    isSelf: boolean;
    wasAdmin: boolean;
    willBeAdmin: boolean;
    willBeActive: boolean;
    remainingActiveAdmins: number;
}): string | null {
    const { isSelf, wasAdmin, willBeAdmin, willBeActive, remainingActiveAdmins } = options;

    if (isSelf && !willBeActive) return 'You cannot deactivate your own account.';
    if (isSelf && wasAdmin && !willBeAdmin) return 'You cannot remove your own administrator role.';

    const losesAdmin = wasAdmin && (!willBeAdmin || !willBeActive);
    if (losesAdmin && remainingActiveAdmins === 0) {
        return "C'est le dernier administrateur actif : le retirer laisserait Zanshin sans personne pour administrer.";
    }
    return null;
}

/** Likewise for deletion, whose consequences are the same but worse. */
export function refuseDeletion(options: { isSelf: boolean; isAdmin: boolean; remainingActiveAdmins: number }): string | null {
    if (options.isSelf) return 'Vous ne pouvez pas supprimer votre propre compte.';
    if (options.isAdmin && options.remainingActiveAdmins === 0) {
        return "C'est le dernier administrateur actif : le supprimer laisserait Zanshin sans personne pour administrer.";
    }
    return null;
}

File: zanshin/zanshin-common/src/main/java/com/asmolabs/zanshin/common/domain/users/AccountRules.java
package com.asmolabs.zanshin.common.domain.users;

import java.util.Optional;
import java.util.regex.Pattern;

/**
 * The rules that stop an administrator from locking themselves out.
 *
 * <p>Pure, because they are rules and not queries: each describes a situation the UI would
 * happily accept and nobody could come back from. <b>There is no rescue screen in Zanshin</b> —
 * with no active administrator left, recovery means a database session.
 *
 * <p>Refused at the account level rather than the screen's: three tabs open on two accounts
 * would otherwise be enough to empty the administrator list.
 */
public final class AccountRules {

    private AccountRules() {}

    public static final int MINIMUM_PASSWORD_LENGTH = 12;

    private static final Pattern USERNAME = Pattern.compile("^[A-Za-z0-9._-]{2,64}$");

    /** Empty if the name is acceptable, otherwise the message to show. */
    public static Optional<String> validateUsername(String username) {
        if (username == null || username.isEmpty()) {
            return Optional.of("A username is required.");
        }
        if (!USERNAME.matcher(username).matches()) {
            return Optional.of("Invalid username: 2 to 64 characters, letters, digits, \". _ -\".");
        }
        return Optional.empty();
    }

    /**
     * A minimum length, and nothing else.
     *
     * <p>No composition rule. Character-class requirements produce {@code Password1!} and
     * encourage reuse; length is the only constraint whose effect on entropy is real.
     *
     * <p><b>The 72-byte ceiling is gone with bcrypt.</b> It existed because bcrypt silently
     * ignores everything past that, so accepting a longer password would have let someone
     * believe a 90-character passphrase protected them while a third of it was never hashed.
     * Argon2id has no such limit, and refusing long passwords was never the goal.
     */
    public static Optional<String> validatePassword(String password) {
        if (password == null || password.isEmpty()) {
            return Optional.of("A password is required.");
        }
        if (password.length() < MINIMUM_PASSWORD_LENGTH) {
            return Optional.of("The password must be at least " + MINIMUM_PASSWORD_LENGTH + " characters.");
        }
        return Optional.empty();
    }

    /**
     * @param remainingActiveAdmins active administrators <b>other than this account</b>
     */
    public record Change(
            boolean isSelf, boolean wasAdmin, boolean willBeAdmin, boolean willBeActive, int remainingActiveAdmins) {}

    /** Empty if the change is allowed, otherwise why it is refused. */
    public static Optional<String> refuseSelfLockout(Change change) {
        if (change.isSelf() && !change.willBeActive()) {
            return Optional.of("You cannot deactivate your own account.");
        }
        if (change.isSelf() && change.wasAdmin() && !change.willBeAdmin()) {
            return Optional.of("You cannot remove your own administrator role.");
        }

        boolean losesAdmin = change.wasAdmin() && (!change.willBeAdmin() || !change.willBeActive());
        if (losesAdmin && change.remainingActiveAdmins() == 0) {
            return Optional.of("This is the last active administrator: removing them would leave Zanshin "
                    + "with nobody able to administer it.");
        }
        return Optional.empty();
    }

    /** Likewise for deletion, whose consequences are the same but worse. */
    public static Optional<String> refuseDeletion(boolean isSelf, boolean isAdmin, int remainingActiveAdmins) {
        if (isSelf) {
            return Optional.of("You cannot delete your own account.");
        }
        if (isAdmin && remainingActiveAdmins == 0) {
            return Optional.of("This is the last active administrator: deleting them would leave Zanshin "
                    + "with nobody able to administer it.");
        }
        return Optional.empty();
    }
}
